pkg/pki/ssh/sign.go

```diff
@@ -0,0 +1,82 @@
+package ssh
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha512"
+	"io"
+	"io/ioutil"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.sshsig#L81
+type MessageWrapper struct {
+	Namespace     string
+	Reserved      string
+	HashAlgorithm string
+	Hash          string
+}
+
+// https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.sshsig#L34
+type WrappedSig struct {
+	MagicHeader   [6]byte
+	Version       uint32
+	PublicKey     string
+	Namespace     string
+	Reserved      string
+	HashAlgorithm string
+	Signature     string
+}
+
+const (
+	magicHeader   = "SSHSIG"
+	hashAlgorithm = "sha512"
+)
+
+func sign(s ssh.AlgorithmSigner, m io.Reader) (*ssh.Signature, error) {
+	data, err := ioutil.ReadAll(m)
+	if err != nil {
+		return nil, err
+	}
+	hf := sha512.New()
+	if _, err := hf.Write(data); err != nil {
+		return nil, err
+	}
+	mh := hf.Sum(nil)
+
+	sp := MessageWrapper{
+		Namespace:     "file",
+		HashAlgorithm: hashAlgorithm,
+		Hash:          string(mh),
+	}
+
+	dataMessageWrapper := ssh.Marshal(sp)
+	dataMessageWrapper = append([]byte(magicHeader), dataMessageWrapper...)
+
+	sig, err := s.SignWithAlgorithm(rand.Reader, dataMessageWrapper, ssh.SigAlgoRSASHA2512)
+	if err != nil {
+		return nil, err
+	}
+	return sig, nil
+}
+
+func Sign(sshPrivateKey string, data []byte) (string, error) {
```

Comment:
can you make `data` an `io.Reader` and just pass it through to the `sign()` call?